### PR TITLE
Create a basic classifier as a proof of concept

### PR DESCRIPTION
**Pull request by ben-heil**
Changes:
- Added the base logic used for training models to classify sepsis data in `classifier.py`
- Set up bread and butter repo files like `util.py`, `model.py`, and `dataset.py`
- Created test cases for some of the functions in `util.py`
- Made `microarray_rnaseq_gene_intersection.py` and `microarray_genes.py` to find which genes can be used in the model

Questions:
Am I doing pytorch right? I structured the script according to my understanding of how models are typically trained in pytorch, but ended up having to do some manual datatype casting. Am I missing anything?

### CODE REVIEW

File: whistl/classifier.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+               'baseline': baseline}
+
+    for epoch in range(num_epochs):
+        train_loss = 0
+        train_correct = 0
+        # Set training mode
+        classifier = classifier.train()
+        for batch in train_loader:
+            expression, labels, ids = batch
+            expression = expression.to(device)
```
Comment (jjc2718) [inline whistl/classifier.py:138]: You may need to pass `device` into this function as an argument - I don't think it's currently defined in the scope of this function (although it's in the code at the bottom of the file).

File: whistl/classifier.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            loss.backward()
+            optimizer.step()
+
+        # Disable the gradient and switch into model evaluation mode
+        with torch.no_grad():
+            classifier = classifier.eval()
+
+            tune_loss = 0
+            tune_correct = 0
+            for tune_batch in tune_loader:
```
Comment (jjc2718) [inline whistl/classifier.py:162]: If you want, I think you can do your model evaluation without needing to use a `DataLoader` (my understanding is they're mostly useful when you need to shuffle data and split it into batches; i.e. during model training).

This might simplify a bit of the logic here, since you can just call something like:
```python
tune_output = classifier(all_tune_expression)
```
without needing to loop over batches. What you have seems correct to me though, so it's up to you which approach you find clearer.

File: whistl/classifier.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+                tune_output = classifier(expression)
+
+                loss = loss_function(tune_output, tune_labels)
+                tune_loss += float(loss)
+                tune_correct += count_correct(tune_output, tune_labels)
+
+        train_accuracy = train_correct / len(train_dataset)
+        tune_accuracy = tune_correct / len(tune_dataset)
+
+        sys.stderr.write('Epoch {}'.format(epoch) + '\n')
```
Comment (jjc2718) [inline whistl/classifier.py]: I'm personally not a huge fan of writing logging messages like this to `stderr` - sometimes external scripts that call your code might exit on messages to `stderr` (assuming it's an actual error), or you may want to write a wrapper script down the line that distinguishes between logging messages and error messages, etc.

If it were me, I'd either have a verbose flag and print logging messages to `stdout` or a log file when it's included, or use [the Python logging module](https://docs.python.org/3.7/library/logging.html) to do something similar.

Up to you, though: my opinion may be in the minority here, since lots of other libraries use `stderr` in the same way you are.

File: whistl/classifier.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+    sys.stderr.write('Generating training dataset...\n')
+    train_dataset = dataset.ExpressionDataset(train_dirs, sample_to_label, label_to_encoding,
+                                              args.gene_file)
+    train_loader = DataLoader(train_dataset, batch_size=16, shuffle=True, num_workers=2,
+                              pin_memory=True)
+    sys.stderr.write('Generating tuning dataset...\n')
+    tune_dataset = dataset.ExpressionDataset(tune_dirs, sample_to_label, label_to_encoding,
+                                             args.gene_file)
+    tune_loader = DataLoader(tune_dataset, batch_size=16, shuffle=True, num_workers=2,
```
Comment (jjc2718) [inline whistl/classifier.py]: Going off of my comment above about `DataLoaders`: out of curiosity, is there a reason to shuffle your tune data? It seems to me like it would be just another source of noise, but maybe I'm missing something.

It definitely helps to shuffle the training data (otherwise your model will overfit to the sample order, I've had this happen in the past) but I can't think of any reason you'd want to do it for the tuning set.

File: whistl/microarray_genes.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+train_dirs = [f for f in subfiles if 'GSE' in f and os.path.isdir(f)]
+
+subfiles = [os.path.join(test_dir, f) for f in os.listdir(test_dir)]
+# Keep only data directories, not anything else that might be in data_dir
+test_dirs = [f for f in subfiles if 'GSE' in f and os.path.isdir(f)]
+
+df_list = []
+for data_dir in train_dirs:
+    study = data_dir.rstrip('/').split('/')[-1]
+    print(study)
```
Comment (jjc2718) [inline whistl/microarray_genes.py]: Might want a more informative logging message here if you plan on using this script more in the future? (Sounds like it's more of a single-use script though, in that case I'm fine with this)

File: whistl/microarray_genes.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    study_file_name = study + '.tsv'
+    data_file = os.path.join(data_dir, study_file_name)
+    curr_df = pd.read_csv(data_file, sep='\t')
+
+    curr_df = curr_df.set_index('Gene')
+    df_list.append(curr_df)
+
+
+combined_df = pd.concat(df_list, axis=1, join='inner')
+print(len(combined_df.columns))
```
Comment (jjc2718) [inline whistl/microarray_genes.py]: Same as above (fine if single-use script, but a more informative message might be nice)

File: whistl/util.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        The data loader for a ExpressionDataset containing the labels to calculate weights for
+
+    Returns
+    -------
+    weights: dict
+        A dictionary mapping encoded labels to weights
+    '''
+    value_counts, total = get_value_counts(train_loader)
+
+    # Weight labels according to the inverse of their frequency
```
Comment (jjc2718) [inline whistl/util.py:90]: Out of curiosity, how imbalanced are the labels for this dataset from what you've seen?

**Review by jjc2718 — APPROVED** [review]
Good start! I think a lot of the casting is unavoidable, since the way that PyTorch deals with data internally is often different from the way that native Python or Numpy does.

Other than my questions in the review (which are mostly stylistic or just for my own understanding), your code looks good to me.

File: whistl/classifier.py (hunk 1 above)
Comment (ben-heil) [reply whistl/classifier.py:138]: Good catch, I always forget that python scoping works differently from c++ and I can't just expect things to break when I pull code into functions

File: whistl/classifier.py (hunk 2 above)
Comment (ben-heil) [reply whistl/classifier.py:162]: That's a good point, I didn't think about that. For now I'll keep the current logic in the (optimistic) hope that at some point I'll have enough validation data that it takes up a nontrivial amount of GPU memory.

File: whistl/classifier.py (hunk 3 above)
Comment (ben-heil) [reply whistl/classifier.py]: Good call, I went ahead and switched over to `logging`

File: whistl/classifier.py (hunk 4 above)
Comment (ben-heil) [reply whistl/classifier.py]: You're right, I had just copied my `train_variable`, shuffling doesn't matter in the tune data

File: whistl/microarray_genes.py (hunk 6 above)
Comment (ben-heil) [reply whistl/microarray_genes.py]: Both of these print statements were something I used to debug an issue and forgot to remove, thanks for pointing them out

File: whistl/util.py (hunk 7 above)
Comment (ben-heil) [reply whistl/util.py:90]: It's roughly a 90/10 split of sepsis:healthy expression
